src/brucke_health_check.erl

```diff
@@ -72,52 +87,72 @@ get_all_routes(true, Status) ->
 get_all_routes(_, Status) -> {[], Status}.
 
 %% @private
+-spec get_working_routes([route()], [{upstream(), pid()}], boolean()) ->
+        {[route_status()], boolean()}.
 get_working_routes(AllRoutes, AllRouteSups, Status) ->
   lists:foldl(
     fun(#route{upstream = Upstream, downstream = Downstream}, {Acc, UStatus}) ->
       {Route, Status1} = check_route(Upstream, AllRouteSups, UStatus),
       {UpstreamId, UpstreamTopic} = Upstream,
       {DownstreamId, DownstreamTopic} = Downstream,
-      Acc1 = [Route#{?UPSTREAM => #{?CLIENT_ID => UpstreamId, ?TOPIC => UpstreamTopic},
-        ?DOWNSTREAM => #{?CLIENT_ID => DownstreamId, ?TOPIC => DownstreamTopic}} | Acc],
+      Acc1 = [Route#{ ?UPSTREAM => #{?CLIENT_ID => UpstreamId,
+                                     ?TOPIC => UpstreamTopic},
+                      ?DOWNSTREAM => #{?CLIENT_ID => DownstreamId,
+                                       ?TOPIC => DownstreamTopic}} | Acc],
       {Acc1, Status1}
     end, {[], Status}, AllRoutes).
 
-%% @private
+%% @private Just one discarded route is enough to toggle status to false.
+-spec apply_discarded([route_status()], [route()], boolean()) -> [route_status()].
 apply_discarded(Working, DiscardedRoutes, Status) ->
-  lists:foldl(  % at least one discarded route is enough to toggle status to false.
-    fun(#route{upstream = {Id, Topic}, reason = Reason}, {Acc, _}) ->
-      {[#{?UPSTREAM => #{?CLIENT_ID => Id, ?TOPIC => Topic},
-        ?STATUS => Reason, ?MEMBERS => []} | Acc], false}
+  lists:foldl(
+    fun(#route{upstream = {UpstreamClientId, UpstreamTopic},
+               reason = Reason
+              }, {Acc, _}) ->
+      {[#{?UPSTREAM => #{?CLIENT_ID => UpstreamClientId,
+                         ?TOPIC => UpstreamTopic},
+          ?STATUS => Reason,
+          ?MEMBERS => []} | Acc], false}
     end, {Working, Status}, DiscardedRoutes).
 
 %% @private
+-spec check_route(upstream(), [{upstream(), pid() | atom()}], boolean()) ->
+        {route_status(), boolean()}.
 check_route(Upstream, Routes, Status) ->
   Res = lists:keyfind(Upstream, 1, Routes),
   check_process_status(Res, Status).
 
 %% @private
--spec check_process_status(pid(), boolean()) -> {map(), boolean()}.
-check_process_status({_, RoutePid, _, _}, Status) when is_pid(RoutePid) ->
+-spec check_process_status(pid(), boolean()) ->
+        {route_status(), boolean()}.
+check_process_status({_, RoutePid}, Status) when is_pid(RoutePid) ->
   try
-    Members = get_all_brucke_members(RoutePid, Status),
-    {#{?STATUS => ok, ?MEMBERS => Members}, Status}
+    {Members, NewStatus} = get_all_brucke_members(RoutePid, Status),
```

Comment:
This was the bug I wanted to fix.
But eventually ended up adding all the specs.